Add support for "*" in zone automation trigger

gravesjohnr opened the pull request:
**Description:**
This change lets you setup automation for ALL zones for a given device.  The zone parameter needs to be "zone.any".  Then, whenever the zone changes, enter and leave automations can be triggered.

In my sample, I use this to run a script whenever my phone enters or leaves ANY zone.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1753

**Example entry for `configuration.yaml` (if applicable):**
```yaml
shell_command:
  send_hangout_message: /home/me/hangouts/sendMessage.sh '{{text}}'

automation:
  - alias: 'Zone Enter'
    trigger:
      platform: zone
      event: enter
      zone: zone.any
      entity_id: device_tracker.owntracks_mys7
    action:
      service: shell_command.send_hangout_message
      data_template:
        text: 'Dad is at {{trigger.to_state.state}}'
  - alias: 'Zone Leave'
    trigger:
      platform: zone
      event: leave
      zone: zone.any
      entity_id: device_tracker.owntracks_mys7
    action:
      service: shell_command.send_hangout_message
      data_template:
        text: 'Dad is leaving {{trigger.from_state.state}}'
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [X ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

gravesjohnr committed:
Added support for a "zone.any" configuration.  This will trigger an automation for an enter or leave for ALL zones given a device.

Sample config:

shell_command:
  send_hangout_message: /home/me/hangouts/sendMessage.sh '{{text}}'

automation:
  - alias: 'Zone Enter'
    trigger:
      platform: zone
      event: enter
      zone: zone.any
      entity_id: device_tracker.owntracks_mys7
    action:
      service: shell_command.send_hangout_message
      data_template:
        text: 'Dad is at {{trigger.to_state.state}}'
  - alias: 'Zone Leave'
    trigger:
      platform: zone
      event: leave
      zone: zone.any
      entity_id: device_tracker.owntracks_mys7
    action:
      service: shell_command.send_hangout_message
      data_template:
        text: 'Dad is leaving {{trigger.from_state.state}}'

robbiet480 commented on homeassistant/components/automation/zone.py:
```diff
@@ -35,17 +35,36 @@ def async_trigger(hass, config, action):
     @callback
     def zone_automation_listener(entity, from_s, to_s):
         """Listen for state changes and calls action."""
+
```
Why the extra lines?

gravesjohnr replied:
Left over from some debugging I was doing.  I'll take them out.

pvizeli commented:
Why not only 'any'?

balloob commented:
I think that you can already achieve it today by using a state automation that is triggered when a state changes from "not_home". No need to extend the support.

balloob commented:
Also, for wildcard we always use `*` (`homeassistant.const.MATCH_ALL`)

gravesjohnr commented:
I used zone.any to avoid having to change the code that checks for valid entries.  The valid entry requires two values separated by a dot.

Balloob, if there is another way, please show sample config that would work with existing code.  Thanks.

gravesjohnr committed:
Removed spaces left over from debug.

Used the constant variable STATE_NOT_HOME rather than hard-coded value of "not_home"

gravesjohnr committed:
Fixed "line too long" problem.

nordlead2005 commented:
What I believe balloob is getting at is looking for "not_home" in the state change.

So something like
```
- alias: Enter Zone
  trigger:
    platform: state
    entity_id: device_tracker.owntracks_mys7
    from: "not_home"
```

you would use "to" instead of "from" for Leave Zone.

gravesjohnr commented:
nordlead2005,  Thanks.  Let me give that a try.

gravesjohnr commented:
I tried using balloob's suggestion: "... use * (homeassistant.const.MATCH_ALL)"

However, using zone.* rather than zone.any makes the entity id pattern fail.  It is expecting alpha-numeric:

From core.py:
# Pattern for validating entity IDs (format: <domain>.<entity>)
ENTITY_ID_PATTERN = re.compile(r"^(\w+)\.(**\w+**)$")

balloob commented:
It should be just `*`, not `zone.*`. And yes, you'll have to update the config validation.

balloob commented:
Will also require tests to be added.

gravesjohnr commented:
balloob.  No problems adding the tests.

I'm wondering about the modifications to the core.py config validation.  Changes I would make there could cause problems in many other modules.  The if/then/else code to handle this particular case would be quite ugly.  All entity IDs are in the format domain.object.  This would be the only case where an entity ID would not be in that format.  zone.* would be cleaner.  I would only need to modify the regex to accept * as well as alphanumeric.

gravesjohnr committed:
Changed to use zone.* rather than zone.any.
Used the constant MATCH_ALL

I modified the config check to allow for the MATCH_ALL value.

gravesjohnr commented:
I've updated the core.py to allow for a * in the 'object' portion of the domain.object entity id format.  This could be useful for other scenarios too.  I've also made changes to use the MATCH_ALL constant.

pvizeli commented:
Don't change entity_id format. Use vol here: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/automation/zone.py#L23

you can make a new check on helpers/config_validation.py like event_entity_id to convert * to MATCH_ALL

balloob commented:
No need to add it to generic config validation, just add '*' support to the validation in zone.py. Don't update other code as you are impacting the whole system.

gravesjohnr committed:
Updated to support "*" configuration and removed changes to core.py

gravesjohnr commented:
I've made the suggested changes to config_validation, removed changes from core.py and made this work with "*" rather than zone.* or zone.any.

The only thing is, in the configuration, the * needs to be in quotes or the yaml parsing doesn't seem to work.  Any suggestions on this or is "*" ok?

e.g. 
automation:
  - alias: 'Zone Enter'
    trigger:
      platform: zone
      event: enter
      **zone: "*"**
      entity_id: device_tracker.owntracks_mys7
    action:
      service: shell_command.send_hangout_message
      data_template:
        text: 'John is at {{trigger.to_state.state}}'

balloob commented on homeassistant/components/automation/zone.py:
```diff
@@ -20,7 +21,7 @@
     vol.Required(CONF_PLATFORM): 'zone',
     vol.Required(CONF_ENTITY_ID): cv.entity_ids,
-    vol.Required(CONF_ZONE): cv.entity_id,
+    vol.Required(CONF_ZONE): cv.zone_id,
```
There is no need to add this to `config_validation.py` as this is not something that will most likely to be reused. Just add this to `automation/zone.py` instead:

```python
vol.Required(CONF_ZONE): vol.Any('*', cv.entity_id)
```

balloob commented on homeassistant/components/automation/zone.py:
```diff
@@ -46,6 +47,23 @@ def zone_automation_listener(entity, from_s, to_s):
+        #
+        #  If zone is "zone.*", trigger automation when entering or leaving
+        #
+        if zone_entity_id == MATCH_ALL and from_s.state != to_s.state:
```
Instead of copy pasting the if-statement, integrate it with the other statement. Extract the check into a variable like done for `from_match` and `to_match`.

gravesjohnr replied:
I have gone over this code several times trying to get this complex if/then/else logic to combine the two scenarios, but it gets SO ugly.  I am happy for suggestions.  It was already so complex that the from_match/to_match was externalised with condition.zone.  When adding a check for zone * AND moving from a 'non-zone' to a 'zone' scenario, it makes the code very unreadable.  Again, I'm happy to see an example if/then/else which is clear and handles the two scenarios.

balloob replied:
ok, please still add a test to verify that this code works.

balloob reviewed (changes requested):
Besides comments this will also require tests to be added for this functionality.

gravesjohnr committed:
Isolated the zone validation to avoid changes to the config_validation code.

homeassistant commented:
Hi @gravesjohnr, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5268).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @gravesjohnr, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5268).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @gravesjohnr, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5268).

Once you do that we will be able to review and accept this pull request.

Thanks!

gravesjohnr committed:
Modified comment to correctly reflect code.

homeassistant commented:
Hi @gravesjohnr, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5268).

Once you do that we will be able to review and accept this pull request.

Thanks!

andrey-git commented:
@gravesjohnr any updates on this PR?

pvizeli commented:
Please reopen this PR if you want finisch that.

gravesjohnr commented:
I've had nothing but problems trying to get the tox environment working.  I finally gave up.

CiquattroFPV commented:
it's closed this PR?
